LeaderboardBackend/Services/Impl/EmailSender.cs

```diff
@@ -0,0 +1,82 @@
+using MailKit.Net.Smtp;
+using Microsoft.Extensions.Options;
+using MimeKit;
+
+namespace LeaderboardBackend.Services;
+
+public class EmailSender : IEmailSender
+{
+    private readonly EmailSenderConfig _config;
+    private readonly ILogger<EmailSender> _logger;
+    private readonly ISmtpClient _smtpClient;
+    private readonly MailboxAddress _sender;
+    private readonly SemaphoreSlim _semaphore;
+
+    public EmailSender(IOptions<EmailSenderConfig> config, ILogger<EmailSender> logger, ISmtpClient smtpClient)
+    {
+        _config = config.Value;
+        _logger = logger;
+        _smtpClient = smtpClient;
+        _sender = new MailboxAddress(name: _config.SenderName, address: _config.SenderAddress);
+        _semaphore = new SemaphoreSlim(1);
+    }
+
+    public async Task EnqueueEmailAsync(string recipientAddress, string subject, string htmlMessage)
+    {
+        if (_config.Smtp is null)
+        {
+            _logger.LogError("Can't send email, SMTP configuration is missing");
+            return;
+        }
+
+        MimeMessage message = new();
+        message.From.Add(_sender);
+        message.To.Add(new MailboxAddress(name: null, address: recipientAddress));
+        message.Subject = subject;
+
+        BodyBuilder bodyBuilder = new() { HtmlBody = htmlMessage };
+        message.Body = bodyBuilder.ToMessageBody();
+
+        await _semaphore.WaitAsync();
+        try
+        {
+            await SendEmail(message);
+        }
+        catch (Exception exception)
+        {
+            _logger.LogError(exception, "Caught exception when sending email (subject: \"{EmailSubject}\")", subject);
```

Comment:
Would it be better to log the exception instead of the email subject? If we do that, we just have to make sure the recipient address isn't logged.

Reply:
Alternatively, log nothing and let the caller decide whether to catch the exception and log or not.

Reply:
(the exception is logged here)

I have no idea if the mailing library can display the email in its exceptions. I'm gonna remove the log for now.

Reply:
Uncle Ted is going blind it seems.